configs/configupgrade/upgrade_expr.go

```diff
@@ -614,6 +614,122 @@ Value:
 	return buf.Bytes(), diags
 }
 
+func validateHilAddress(address, filename string) tfdiags.Diagnostics {
+	parts := strings.Split(address, ".")
+	var diags tfdiags.Diagnostics
+
+	label, ok := getResourceLabel(parts)
+	if ok && !hcl2syntax.ValidIdentifier(label) {
+		// We can't get any useful source location out of HIL unfortunately
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			fmt.Sprintf("Invalid address (%s) in ./%s", address, filename),
+			// The label could be invalid for another reason
+			// but this is the most likely, so we add it as hint
+			"Labels of resources and data sources can no longer start with digits."))
+	}
+
+	return diags
+}
+
+func getResourceLabel(parts []string) (string, bool) {
+	if len(parts) < 1 {
+		return "", false
+	}
+
+	if parts[0] == "data" {
+		if len(parts) < 3 {
+			return "", false
+		}
+		return parts[2], true
+	}
+
+	if len(parts) < 2 {
+		return "", false
+	}
+
+	return parts[1], true
```

Comment:
If `parts[0]` were something like `local`, `var`, `module`, etc then I guess it would end up here and return the name of the local value, variable, module, or whatever. Those are also subject to the new rule, so that's fine and correct, but the error message in `validateHilAddress` only talks about resources and data sources, so it may lead to some confusing if there's some other object with a now-invalid name.

Listing all of the possible things that could be named in that message would be quite unwieldy I think, so maybe we could hedge with something like this:

```
Names of objects (resources, modules, etc) may no longer start with digits.
```

AFAICR `data` is the only case where the main name of the thing isn't in the index 1 position, so I think this should be safe, because there's never a case where the required prefix part is shorter than two parts or where the second position allows non-identifiers.

Reply:
Good point - totally agreed.

Reply:
I will merge these two commits together and close #21125 then.